### ISSUE TITLE
`MissingImport` flags non-namespaced classes referenced from files in the global namespace

### ISSUE THREAD

**Issue by lukasbestle** [phpmd/phpmd#910]
- PHPMD version: 2.10.2
- PHP Version: 8.0.9
- Installation type: Composer
- Operating System / Distribution & Version: macOS 10.15.7 & Ubuntu 20.04.3 LTS (GitHub Actions)

### Current Behavior

The `MissingImport` rule expects classes from the global namespace to be explicitly imported even when the file that references the class doesn't have a namespace. If an import is used in this case, PHP triggers the warning:

> The use statement with non-compound name … has no effect

### Expected Behavior

PHPMD shouldn't flag class references when both the referenced class and the referencing file are in the global namespace.

### Steps To Reproduce:

**Example file:**

```php
<?php

function test() {
    $test = new DateInterval();
}
```

**Command:**

```bash
phpmd test.php ansi cleancode
```

**Output:**

```
FILE: /path/to/test.php
---------------------------------
 4 | VIOLATION | Missing class import via use statement (line '4', column '15').

Found 1 violation and 0 errors in 66ms
```

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by kylekatarnls** [comment]
Hum, I feel like this PHP warning is a bit unfortunate as it would have been convenient to keep the list of used classes, so you can add a namespace later without having to search for all the references.

Sure, it would be nicer if we would handle this particular case and ignore MissingImport for such first-level classes in non-namespaced files.

But still, note that we generally assume every PHP file has a namespace, because it's actually also a very good practice. I guess we could even have a new rule for file without namespace because in my opinion, it's a code smell: e.g. code not properly encapsulated.

**Comment by distantnative** [comment]
@kylekatarnls Creating such a rule could be an additional measure. Nevertheless, when such rule is deactivated, phpmd  should still not throw the described warning as it conflicts with PHP's warning. It would be very unfortunate if we end up with a situation where we either have warnings from phpmd or PHP itself. We have a project where giving up that specific non-namespace file won't be an option.

**Comment by lukasbestle** [comment]
@kylekatarnls Thank you for your super quick reply!

> But still, note that we generally assume every PHP file has a namespace, because it's actually also a very good practice.

I fully agree, but unfortunately it's not always possible. The file in question defines helper functions that are to be called from [PHP templates](https://getkirby.com/docs/guide/templates/basics#the-page-variable) in Kirby CMS. In this context our helper functions cannot be namespaced as that would be a huge usability concern for template devs (it would require adding a huge `use` block at the top of each template). That's sadly one of the limitations of otherwise really awesome native PHP templating.

> Hum, I feel like this PHP warning is a bit unfortunate as it would have been convenient to keep the list of used classes, so you can add a namespace later without having to search for all the references.

Yep, I was surprised by the warning as well. I've never seen it before and I agree with you that it shouldn't be there. Unfortunately that's how it is, so IMO we need to fix this case in the `MissingImport` rule.

**Comment by rafaelbernard** [comment]
I had success changing the value for [`ignore-global`](https://phpmd.org/rules/cleancode.html#missingimport). Is this an option for you?

```xml
<rule ref="rulesets/cleancode.xml">
        <exclude name="MissingImport" />
    </rule>
    <rule ref="rulesets/cleancode.xml/MissingImport">
        <properties>
            <property name="ignore-global" value="true"/>
        </properties>
    </rule>
```

**Comment by lukasbestle** [comment]
Thanks for sharing, it's a possible workaround. The disadvantage is that this also disables violations in files that *do* have a namespace declaration (in which the explicit import with `use` really makes sense).

**Comment by bubokko** [comment]
Is there any update on this after almost 4 years? Assuming that we want namespaces in all PHP files is wrong. I got tons of simple CLI scripts that never declare any classes and I don't want them to have a namespace. I can't win with PHP Mess Detector here - if I don't add `use`, it complains. If I add `use`, I get a PHP warning:

> Warning: The use statement with non-compound name 'DateTimeZone' has no effect in /Users/me/www/time/time.php on line 3

This is ridiculous.

**Comment by AJenbo** [comment]
Feel free to provide a pull request

**Comment by bubokko** [comment]
> Feel free to provide a pull request

No, I'm not gonna study the entire thing in order to fix the annoyance, thank you.

**Comment by AJenbo** [comment]
That's ridiculous

**Comment by kylekatarnls** [comment]
If you have files with no namespace, you can use ignore-global option https://phpmd.org/rules/cleancode.html#missingimport
